<?xml version="1.0"?>
<robot xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor"
       xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller"
       xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface">

  <property name="M_PI" value="3.1415926535897931" />
 
  <macro name="m3_l_palm" params="side parent reflect *origin">

  <joint name="palm_joint_left" type="fixed">
    <origin xyz="0 0 -0.04414" rpy="0 0 0" />
    <parent link="${parent}"/>
    <child link="palm_left" />
  </joint>

     <link name="palm_left">
      <inertial>
        <mass value="0.663" />
	<origin xyz="0.001459 0.001619 -0.051312"/>
	<inertia ixx="0.000873"
	iyy="0.000597"
	izz="0.000565"
	ixy="0.000031"
	ixz="-0.000008"
	iyz="0.000016" />
      </inertial>

      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="palm_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_LT8_palm_left.dae" />
	      </geometry>  
      </visual>

      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="palm_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_LT8_palm_left.stl" />
	      </geometry>
      </collision>
    </link>

<!-- Start Thumb -->
    <joint name="left_hand_j0" type="revolute">
      <axis xyz="0 0 1" />
      <origin xyz="-0.02 -0.028 -0.03" rpy="0 0 0" />
      <parent link="palm_left"/>
      <child link="thumb0_left"/>
      <limit effort="100" velocity="${20.0*M_PI/180}" lower="${-40*M_PI/180}" upper="${15*M_PI/180}" />
    </joint>

    <link name="thumb0_left">
      <inertial>
        <mass value="0.0117" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="thumb0_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_RJ0_thumb_abductlink.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="thumb0_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_RJ0_thumb_abductlink.stl" />
	      </geometry>
      </collision>
    </link>
  
    <joint name="left_hand_j1" type="revolute">
      <axis xyz="0 -1 0"/>
     <limit effort="100" velocity="${75*M_PI/180}" lower="${-80.919*M_PI/180}" upper="${80.91*M_PI/180}" />    
      <origin xyz="-0.018 0 0" rpy="0 0 0" />
      <parent link="thumb0_left"/>
      <child link="thumb1_left"/>
    </joint>
    <link name="thumb1_left">
      <inertial>
        <mass value="0.0344" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>
      <visual>
	      <material name="GreyB" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="thumb1_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_thumb_knuckle1.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="thumb1_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_thumb_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="left_hand_j2" type="revolute">
      <axis xyz="0 -1 0" />
      <origin xyz="-0.044 0 0" rpy="0 0 0" />
      <parent link="thumb1_left" />
      <child link="thumb2_left"/>
      <limit effort="100" velocity="${50*M_PI/180}" lower="${-14.94*M_PI/180}" upper="${14.94*M_PI/180}" />    
    </joint>

    <link name="thumb2_left">  
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="thumb_knuckle2_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_thumb_knuckle2.dae" />
	      </geometry>  
      </visual>
      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="thumb_knuckle2_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_thumb_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>
    <joint name="left_hand_j3" type="revolute">
      <axis xyz="0 -1 0" />
      <origin xyz="-0.035 0 0" rpy="0 0 0" />
      <parent link="thumb2_left" />
      <child link="thumb_end_left"/>
      <limit effort="100" velocity="${50*M_PI/180}" lower="${-14.94*M_PI/180}" upper="${14.94*M_PI/180}" />    
    </joint>
    <link name="thumb_end_left"/>
<!-- Start Index Finger -->
    <joint name="left_hand_j4" type="revolute">
      <axis xyz="0 1 0" />
      <origin xyz="0 0.028 -0.101" rpy="0 0 0" />
      <parent link="palm_left" />
      <child link="index0_left"/>
	<limit effort="100" velocity="${50*M_PI/180}" lower="${-7.072*M_PI/180}" upper="${37.072*M_PI/180}" />    
    </joint>

    <link name="index0_left">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="index0_geom_left">		      
		     <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle1.dae" />
	      </geometry>  
      </visual>      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="index0_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="left_hand_j5" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.99*M_PI/180}" upper="${34.99*M_PI/180}" /> 
      <origin xyz="0 0 -0.044" rpy="0 0 0" />
      <parent link="index0_left"/>
      <child link="index1_left"/>
    </joint>

    <link name="index1_left"> 
      <inertial>
        <mass value="0.0429" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>     
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="index1_geom_left">		      
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle2.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="index1_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="left_hand_j6" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0 -0.032" rpy="0 0 0" />
      <parent link="index1_left"/>
      <child link="index2_left"/>
    </joint>

    <link name="index2_left">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="index2_geom_left">		      
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle3.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="index2_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle3.stl" />
	      </geometry>
      </collision>
    </link>
    <joint name="left_hand_j7" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0.008 -0.025" rpy="0 0 0" />
      <parent link="index2_left"/>
      <child link="index_end_left"/>
    </joint>
    <link name="index_end_left"/>
<!-- Start Ring Finger -->
    <joint name="left_hand_j8" type="revolute">
      <axis xyz="0 1 0" />
      <origin xyz="0 0 -0.101" rpy="0 0 0" />
      <parent link="palm_left" />
      <child link="ring0_left"/>
	<limit effort="100" velocity="${50*M_PI/180}" lower="${-7.072*M_PI/180}" upper="${37.072*M_PI/180}" />    
    </joint>

    <link name="ring0_left">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="ring0_geom_left">
		     <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle1.dae" />
	      </geometry>  
      </visual>      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="ring0_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="left_hand_j9" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.99*M_PI/180}" upper="${34.99*M_PI/180}" /> 
      <origin xyz="0 0 -0.044" rpy="0 0 0" />
      <parent link="ring0_left"/>
      <child link="ring1_left"/>
    </joint>

    <link name="ring1_left">    
      <inertial>
        <mass value="0.0429" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>       
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="ring1_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle2.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="ring1_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="left_hand_j10" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0 -0.032" rpy="0 0 0" />
      <parent link="ring1_left"/>
      <child link="ring2_left"/>
    </joint>

    <link name="ring2_left">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="ring2_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle3.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="ring2_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle3.stl" />
	      </geometry>
      </collision>
    </link>
    <joint name="left_hand_j11" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0.008 -0.025" rpy="0 0 0" />
      <parent link="ring2_left"/>
      <child link="ring_end_left"/>
    </joint>
    <link name="ring_end_left"/>
<!-- Start Pinky Finger -->
    <joint name="left_hand_j12" type="revolute">
      <axis xyz="0 1 0" />
      <origin xyz="0 -0.028 -0.101" rpy="0 0 0" />
      <parent link="palm_left" />
      <child link="pinky0_left"/>
	<limit effort="100" velocity="${50*M_PI/180}" lower="${-7.072*M_PI/180}" upper="${37.072*M_PI/180}" />    
    </joint>

    <link name="pinky0_left">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyA" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="pinky0_geom_left">
		     <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle1.dae" />
	      </geometry>  
      </visual>      
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="pinky0_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle1.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="left_hand_j13" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.99*M_PI/180}" upper="${34.99*M_PI/180}" /> 
      <origin xyz="0 0 -0.044" rpy="0 0 0" />
      <parent link="pinky0_left"/>
      <child link="pinky1_left"/>
    </joint>

    <link name="pinky1_left">      
      <inertial>
        <mass value="0.0429" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>     
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="pinky1_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle2.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="pinky1_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle2.stl" />
	      </geometry>
      </collision>
    </link>

    <joint name="left_hand_j14" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0 -0.032" rpy="0 0 0" />
      <parent link="pinky1_left"/>
      <child link="pinky2_left"/>
    </joint>

    <link name="pinky2_left">
      <inertial>
        <mass value="0.01" />
	<origin xyz="0.0 0.0 0.0"/>
	<inertia ixx="0.0"
	iyy="0.0"
	izz="0.0"
	ixy="0.0"
	ixz="0.0"
	iyz="0.0" />
      </inertial>        
      <visual>
	      <material name="GreyC" />
	      <origin xyz="0 0 0" rpy="0 0 0 " />
	      <geometry name="pinky2_geom_left">
		      <mesh filename="package://${MESH}/H2R3/H2R3_finger_knuckle3.dae" />
	      </geometry>  
      </visual>
      <collision>
	      <origin xyz="0 0 0" rpy="0.0 0.0 0.0 " />
	      <geometry name="pinky2_collision_left">
		      <mesh filename="package://${MESH_COLL}/H2R3/H2R3_finger_knuckle3.stl" />
	      </geometry>
      </collision>
    </link>
    <joint name="left_hand_j15" type="revolute">
      <axis xyz="0 1 0" />
      <limit effort="100" velocity="${1600*M_PI/180}" lower="${-34.21*M_PI/180}" upper="${34.213*M_PI/180}" /> 
      <origin xyz="0 0.008 -0.025" rpy="0 0 0" />
      <parent link="pinky2_left"/>
      <child link="pinky_end_left"/>
    </joint>
    <link name="pinky_end_left"/>
</macro>

  <macro name="m3_l_hand" params="side parent reflect *origin">
    <m3_l_palm side="${side}" reflect="${reflect}" parent="${parent}">
      <insert_block name="origin" />
    </m3_l_palm>
  </macro>


</robot>
